contracts/beyondFramework/dca/base/DCABaseUpgradeable.sol

```diff
@@ -0,0 +1,1168 @@
+    uint256 public lastInvestmentTimestamp;
+    uint256 public minDepositAmount;
+
+    uint16 public positionsLimit;
+
+    address[] public depositToBluechipSwapPath;
+    address[] public bluechipToDepositSwapPath;
+
+    BluechipInvestmentState public bluechipInvestmentState;
+
+    InvestQueueLib.InvestQueue private globalInvestQueue;
+    DcaHistoryLib.DcaHistory private dcaHistory;
+    SwapLib.Router public router;
+
+    TokenInfo public emergencyExitDepositToken;
+    uint256 public emergencySellDepositPrice;
+    TokenInfo public emergencyExitBluechipToken;
+    uint256 public emergencySellBluechipPrice;
+
+    mapping(address => DcaDepositor) private depositors;
+
+    uint256[7] private __gap;
+
+    // solhint-disable-next-line
+    function __DCABaseUpgradeable_init(DcaStrategyInitArgs calldata args)
+        internal
+        onlyInitializing
+    {
+        __Ownable_init();
+        __Pausable_init();
+        __ReentrancyGuard_init();
+
+        setBluechipInvestmentState(BluechipInvestmentState.Investing);
+        setDepositFee(args.depositFee);
+        setDcaInvestor(args.dcaInvestor);
+        setDepositTokenInto(args.depositTokenInfo);
+        setInvestmentPeriod(args.investmentPeriod);
+        setLastInvestmentTimestamp(args.lastInvestmentTimestamp);
+        setMinDepositAmount(args.minDepositAmount);
+        setPositionsLimit(args.positionsLimit);
+        setRouter(args.router);
+        setSwapPath(
+            args.depositToBluechipSwapPath,
+            args.bluechipToDepositSwapPath
+        );
+    }
+
+    modifier onlyDcaInvestor() {
+        require(_msgSender() == dcaInvestor, "Unauthorized");
+        _;
+    }
+
+    modifier onlyPortfolio() {
+        bool authorized;
+        for (uint256 i = 0; i < portfolios.length; i++) {
+            if (portfolios[i] == _msgSender()) {
+                authorized = true;
+                break;
+            }
+        }
+
+        require(authorized, "Unauthorized");
+        _;
+    }
+
+    modifier nonEmergencyExited() {
+        require(
+            bluechipInvestmentState != BluechipInvestmentState.EmergencyExited,
+            "Strategy is emergency exited"
+        );
+
+        _;
+    }
+
+    modifier emergencyWithdrawOnEmergencyExitedStatus(address sender) {
+        // if emergency exited then user should receive everything in deposit asset
+        if (isEmergencyExited()) {
+            _emergencyWithdrawUserDeposit(sender);
+            return;
+        }
+
+        _;
+    }
+
+    receive() external payable {}
+
+    // ----- Base Class Methods -----
+    function deposit(uint256 amount, uint8 amountSplit)
+        public
+        virtual
+        nonReentrant
+        whenNotPaused
+        nonEmergencyExited
+    {
+        _deposit(_msgSender(), amount, amountSplit);
+    }
+
+    function depositFor(
+        address sender,
+        uint256 amount,
+        uint8 amountSplit
+    )
+        public
+        virtual
+        onlyPortfolio
```

Comment:
I think this would be useful for everyone, let's say I want to deposit for you. Why do we only allow for portfolios?

Reply:
actually allowing anyone could be a bad idea: let's say you invested 1 million, and then I keep investing minimum amount like a thousand times, and then hitting the period limit. This means next time, you would need to withdraw everything, before investing which might potentially cause big losses to you.